File: client/src/models/BoardConfiguration.ts
export class PointConfiguration {
    player1: number;
    player2: number;

    constructor(player1: number, player2: number) {
        this.player1 = player1;
        this.player2 = player2;
    }
}

// Starting configuration for the board
const DEFAULT_POINTS: PointConfiguration[] = [
    new PointConfiguration(0, 2), // Point 1
    new PointConfiguration(0, 0), // Point 2
    new PointConfiguration(0, 0), // Point 3
    new PointConfiguration(0, 0), // Point 4
    new PointConfiguration(0, 0), // Point 5
    new PointConfiguration(5, 0), // Point 6
    new PointConfiguration(0, 0), // Point 7
    new PointConfiguration(3, 0), // Point 8
    new PointConfiguration(0, 0), // Point 9
    new PointConfiguration(0, 0), // Point 10
    new PointConfiguration(0, 0), // Point 11
    new PointConfiguration(0, 5), // Point 12
    new PointConfiguration(5, 0), // Point 13
    new PointConfiguration(0, 0), // Point 14
    new PointConfiguration(0, 0), // Point 15
    new PointConfiguration(0, 0), // Point 16
    new PointConfiguration(0, 3), // Point 17
    new PointConfiguration(0, 0), // Point 18
    new PointConfiguration(0, 5), // Point 19
    new PointConfiguration(0, 0), // Point 20
    new PointConfiguration(0, 0), // Point 21
    new PointConfiguration(0, 0), // Point 22
    new PointConfiguration(0, 0), // Point 23
    new PointConfiguration(2, 0)  // Point 24
];

export class BoardConfiguration {
    points: PointConfiguration[];
    bar: PointConfiguration;

    constructor(points?: PointConfiguration[], bar?: PointConfiguration) {
        this.points = points ?? DEFAULT_POINTS;
        this.bar = bar ?? new PointConfiguration(0, 0);
    }
}

export class Match {
    _id: number;
    player1: string;
    player2: string;
    boardConfiguration: BoardConfiguration;
    dice: {
        roll: number[];
        used: number[];
    };
    turn: number;
    created_at: Date;
    updated_at: Date;
    status: string;

    constructor(_id: number, player1: string, player2: string, boardConfiguration: BoardConfiguration, dice: { roll: number[]; used: number[]; }, turn: number, created_at: Date, updated_at: Date, status: string) {
        this._id = _id;
        this.player1 = player1;
        this.player2 = player2;
        this.boardConfiguration = boardConfiguration ?? new BoardConfiguration();
        this.dice = dice;
        this.turn = turn;
        this.created_at = created_at;
        this.updated_at = updated_at;
        this.status = status;
    }
}

File: server/models/board_configuration.py
from fastapi.security import OAuth2PasswordBearer

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")
from typing import List

from pydantic import BaseModel, Field
from services.database import default_id


class Point(BaseModel):
    player1: int
    player2: int

    def __init__(self, player1: int = 0, player2: int = 0):
        super().__init__(player1=player1, player2=player2)


# Starting configuration of the board
DEFAULT_POINTS: List[Point] = [
    Point(0, 2),  # Point 1
    Point(0, 0),  # Point 2
    Point(0, 0),  # Point 3
    Point(0, 0),  # Point 4
    Point(0, 0),  # Point 5
    Point(5, 0),  # Point 6
    Point(0, 0),  # Point 7
    Point(3, 0),  # Point 8
    Point(0, 0),  # Point 9
    Point(0, 0),  # Point 10
    Point(0, 0),  # Point 11
    Point(0, 5),  # Point 12
    Point(5, 0),  # Point 13
    Point(0, 0),  # Point 14
    Point(0, 0),  # Point 15
    Point(0, 0),  # Point 16
    Point(0, 3),  # Point 17
    Point(0, 0),  # Point 18
    Point(0, 5),  # Point 19
    Point(0, 0),  # Point 20
    Point(0, 0),  # Point 21
    Point(0, 0),  # Point 22
    Point(0, 0),  # Point 23
    Point(2, 0)  # Point 24
]


class BoardConfiguration(BaseModel):
    points: List[Point] = DEFAULT_POINTS
    bar: Point = Point(player1=0, player2=0)

    def __init__(self, points: List[Point] = DEFAULT_POINTS, bar: Point = Point(player1=0, player2=0)):
        super().__init__(points=points, bar=bar)


class Match(BaseModel):
    id: str = Field(default_factory=default_id, alias="_id")
    player1: str
    player2: str
    board_configuration: BoardConfiguration = BoardConfiguration()
    dice: List[int] = []
    available: List[int] = []
    turn: int = 0
    status: str = "pending"
    rounds_to_win: int
    winsP1: int = 0
    winsP2: int = 0


class CreateInviteRequest(BaseModel):
    opponent_username: str
    rounds_to_win: int


class AcceptInviteRequest(BaseModel):
    invite_id: str
